### PR TITLE
Add payroll app to all templates

### PR DESCRIPTION
**Pull request by ajsantander**
:construction:  WIP :construction: 

#### Tasks
- [x] Add Payroll app functionality in base template
- [x] Add Payroll app in company template (review before applying to all others)
- [x] Company board template
- [x] Membership template
- [x] Reputation template

### CODE REVIEW

File: shared/helpers/abi.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    })
+  },
+
+  encodeFunctionCall: (funcSignature, types, params) => {
+    params = abiUtil.stringifyParams(params)
+    return abi.encodeFunctionSignature(funcSignature) + abi.encodeParameters(types, params).replace('0x', '');
+  }
+}
+
+module.exports = abiUtil
```
Comment (facuspagnuolo) [inline shared/helpers/abi.js]: Let's inline `abiUtil` to be consistent with the rest of the modules here (see other module exports for reference).

File: shared/helpers/abi.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+          return p
+        }
+        else {
+          return `${p}`
+        }
+      }
+    })
+  },
+
+  encodeFunctionCall: (funcSignature, types, params) => {
```
Comment (facuspagnuolo) [inline shared/helpers/abi.js]: We are following this convention

```suggestion
  encodeFunctionCall(signature, types, params) {
```

File: shared/scripts/test-ganache.sh [hunk 3]
```diff
@@ -19,7 +19,7 @@ cleanup() {
 setup_testing_variables() {
   PORT=${PORT-8545}
   BALANCE=${BALANCE-100000}
-  GAS_LIMIT=${GAS_LIMIT-8000000}
+  GAS_LIMIT=${GAS_LIMIT-10000000}
```
Comment (facuspagnuolo) [inline shared/scripts/test-ganache.sh]: We shouldn't be working with non-real gas values. If we are hitting a gas limit, we should split the transactions

**Comment by facuspagnuolo** [comment]
Left a few comments, would you mind updating this branch with `master`?

File: shared/contracts/BaseTemplate.sol [hunk 4]
```diff
@@ -128,6 +132,25 @@ contract BaseTemplate is APMNamehash, IsContract {
...
+    function _installPayrollApp(Kernel _dao, Finance _finance, address _denominationToken, IFeed _priceFeed, uint64 _rateExpiryTime) internal returns (Payroll) {
+        Payroll payroll = Payroll(_installNonDefaultApp(_dao, PAYROLL_APP_ID));
+        payroll.initialize(_finance, _denominationToken, _priceFeed, _rateExpiryTime);
+        return payroll;
+    }
+
+    function _createPayrollPermissions(ACL _acl, Payroll _payroll, address _grantee, address _manager) internal {
+        _acl.createPermission(_grantee, _payroll, _payroll.ADD_BONUS_ROLE(), _manager);
+        _acl.createPermission(_grantee, _payroll, _payroll.ADD_EMPLOYEE_ROLE(), _manager);
+        _acl.createPermission(_grantee, _payroll, _payroll.ADD_REIMBURSEMENT_ROLE(), _manager);
```
Comment (sohkai) [inline shared/contracts/BaseTemplate.sol]: We may want a different manager for the `ADD_BONUS`, `ADD_REIMBURSEMENT`, and `ADD_EMPLOYEE` roles.

The other ones are very powerful “setting” management roles, whereas these are directly targeted at managing employees. Perhaps we can have the concept of the “Payroll settings manager” (should be Voting or etc.) and the “Payroll employee manager” (could be a high privilege account)?

File: shared/contracts/BaseTemplate.sol [hunk 5]
```diff
@@ -128,6 +132,26 @@ contract BaseTemplate is APMNamehash, IsContract {
...
+    /* PAYROLL */
+
+    function _installPayrollApp(Kernel _dao, Finance _finance, address _denominationToken, IFeed _priceFeed, uint64 _rateExpiryTime) internal returns (Payroll) {
+        Payroll payroll = Payroll(_installNonDefaultApp(_dao, PAYROLL_APP_ID));
+        payroll.initialize(_finance, _denominationToken, _priceFeed, _rateExpiryTime);
+        return payroll;
+    }
+
+    function _createPayrollPermissions(ACL _acl, Payroll _payroll, address _employeeManager, address _settingsManager, address _permissionsManager) internal {
+        _acl.createPermission(_employeeManager, _payroll, _payroll.ADD_BONUS_ROLE(), _permissionsManager);
```
Comment (sohkai) [inline shared/contracts/BaseTemplate.sol]: Let's leave some comments for this separation of roles

File: shared/contracts/BaseTemplate.sol [hunk 6]
```diff
@@ -257,4 +281,12 @@ contract BaseTemplate is APMNamehash, IsContract {
     function _ensureAragonIdIsValid(address _aragonID) internal view {
         require(isContract(address(_aragonID)), ERROR_ARAGON_ID_NOT_CONTRACT);
     }
+
+    /* HELPERS */
+
+    // TODO: Move to helper (in this repo for now)
+    function _toAddress(uint256 _value) internal pure returns(address) {
+        require(_value <= uint160(-1), ERROR_CANNOT_CAST_VALUE_TO_ADDRESS);
+        return address(_value);
```
Comment (sohkai) [inline shared/contracts/BaseTemplate.sol]: Does solidity overflow if the value is too big?

File: shared/helpers/abi.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+        return p
+      }
+      else {
+        return `${p}`
+      }
+    }
+  })
+}
+
+function encodeFunctionCall(funcSignature, types, params) {
```
Comment (sohkai) [inline shared/helpers/abi.js]: Can we not just use `web3-eth-abi`'s [`encodeFunctionCall()`](https://github.com/ethereum/web3.js/blob/1.x/packages/web3-eth-abi/src/index.js#L200) directly?

File: templates/trust/package.json [hunk 8]
```diff
@@ -43,6 +43,7 @@
     "@aragon/apps-agent": "^1.0.0",
     "@aragon/apps-vault": "^4.1.0",
     "@aragon/apps-voting": "^2.1.0",
+    "@aragon/apps-payroll": "^1.0.0-rc.1",
```
Comment (sohkai) [inline templates/trust/package.json]: Don't think we need this dependency here 😄

File: shared/helpers/assertRevert.js [hunk 9]
```diff
@@ -27,8 +27,8 @@ module.exports = web3 => {
     }
   }
 
-  async function assertRevertGeth(request, reason) {
-    const tx = request.params[0]
+  async function assertRevertGeth(truffleInstance, requestOrTx, reason) {
```
Comment (sohkai) [inline shared/helpers/assertRevert.js]: It seems quite odd to require the truffle-contract instance here (and it deviates from a lot of other `assertRevert` code we have across the org), why was this needed?

File: templates/company/test/company.js [hunk 10]
```diff
@@ -105,7 +123,15 @@ contract('Company', ([_, owner, holder1, holder2]) => {
...
             } else if (creationStyle === 'separate') {
               tokenReceipt = await template.newToken(TOKEN_NAME, TOKEN_SYMBOL, { from: owner })
-              instanceReceipt = await template.newInstance(daoID, HOLDERS, STAKES, VOTING_SETTINGS, DEFAULT_FINANCE_PERIOD, useAgentAsVault, { from: owner })
+              if (installPayroll) {
+                const dummyPayrollFeed = template.address
+                const employeeManager = useOwnerAsEmployeeManager ? owner : ZERO_ADDRESS
+                const payrollSettings = [PAYROLL_DENOMINATION_TOKEN, dummyPayrollFeed, PAYROLL_RATE_EXPIRY_TIME, employeeManager]
+                instanceReceipt = await newInstance(daoID, HOLDERS, STAKES, VOTING_SETTINGS, DEFAULT_FINANCE_PERIOD, useAgentAsVault, payrollSettings)
+              }
+              else {
```
Comment (sohkai) [inline templates/company/test/company.js]: Super style nitpick, we generally put `else`s on the trailing bracket

Prefer

```js
if {
} else {
}
```

Over

```js
if {
}
else {
}
```

File: templates/company/test/company.js [hunk 11]
```diff
@@ -243,6 +277,63 @@ contract('Company', ([_, owner, holder1, holder2]) => {
...
+        if (creationStyle === 'separate') {
+          context('when installing the payroll', () => {
+
+            context('when using the voting app for employee management', () => {
+              itHandlesInstanceCreationsProperly(true, true, false)
+
+              it('should have payroll app correctly setup', async () => {
+                assert.isTrue(await payroll.hasInitialized(), 'payroll not initialized')
+                assert.equal(await payroll.denominationToken(), PAYROLL_DENOMINATION_TOKEN)
+                assert.equal(await payroll.feed(), template.address)
```
Comment (sohkai) [inline templates/company/test/company.js]: Perhaps we could use a constant for the feed address too, in the relevant contexts.

File: templates/company/contracts/CompanyTemplate.sol [hunk 12]
```diff
@@ -48,24 +52,64 @@ contract CompanyTemplate is BaseTemplate {
...
 
     function newInstance(string _id, address[] _holders, uint256[] _stakes, uint64[3] _votingSettings, uint64 _financePeriod, bool _useAgentAsVault) public {
+        _verifyCompanyParameters(_holders, _stakes, _votingSettings);
+        (Kernel dao, ACL acl) = _createDAO();
+        (, Voting voting) = _setupApps(dao, acl, _holders, _stakes, _votingSettings, _financePeriod, _useAgentAsVault);
+        _transferRootPermissionsFromTemplate(dao, voting);
+        _registerID(_id, dao);
+    }
+
+    function newInstance(
```
Comment (sohkai) [inline templates/company/contracts/CompanyTemplate.sol]: We should add a comment to `newInstanceAndToken()` noting that we can't fit the payroll app in a single block with it (I'm seeing that at 6.9m and the delta for including the Payroll app at 1.1m?)

File: templates/company/contracts/CompanyTemplate.sol [hunk 13]
```diff
@@ -48,24 +52,64 @@ contract CompanyTemplate is BaseTemplate {
...
-        _mintTokens(acl, tokenManager, _holders, _stakes);
+        return (finance, voting);
+    }
 
-        _setupPermissions(dao, acl, agentOrVault, voting, finance, tokenManager, _useAgentAsVault);
+    function _setupPayrollApp(Kernel _dao, ACL _acl, uint256[4] _payrollSettings, Finance _finance, Voting _voting) internal {
+        address denominationToken = _toAddress(_payrollSettings[0]);
+        IFeed priceFeed = IFeed(_toAddress(_payrollSettings[1]));
+        uint64 rateExpiryTime = _payrollSettings[2].toUint64();
+        address employeeManager = _toAddress(_payrollSettings[3]);
```
Comment (sohkai) [inline templates/company/contracts/CompanyTemplate.sol]: One small thing is perhaps doing this coercion at the start of `newInstance()` instead, so the user doesn't waste gas creating an entire org before their input fails here.

File: shared/contracts/BaseTemplate.sol (hunk 6 above)
Comment (facuspagnuolo) [reply shared/contracts/BaseTemplate.sol]: yup

File: templates/trust/package.json (hunk 8 above)
Comment (ajsantander) [reply templates/trust/package.json]: There's something weird going on, if that dependency isn't there, Trust doesn't compile. Or at least that was the case with the CI.

File: shared/helpers/assertRevert.js (hunk 9 above)
Comment (ajsantander) [reply shared/helpers/assertRevert.js]: For some odd reason, extracting the tx and sending it via web3.eth.sendTransaction does not produce a revert error with the correct reason. Sending it via the truffle object does not have that problem.

File: templates/company/test/company.js (hunk 11 above)
Comment (ajsantander) [reply templates/company/test/company.js]: Sure, but beware that it needs to be a contract since `isContract()` is run on that address.

File: shared/helpers/assertRevert.js (hunk 9 above)
Comment (sohkai) [reply shared/helpers/assertRevert.js]: Hmm, this is awkward. It reverts with a different reason when sending through with web3?

Comment (ajsantander) [reply shared/helpers/assertRevert.js]: It reverts with no reason, just a plain "reverted" error.

File: templates/company/contracts/CompanyTemplate.sol (hunk 13 above)
Comment (facuspagnuolo) [reply templates/company/contracts/CompanyTemplate.sol]: see my summary comment bellow

**Comment by facuspagnuolo** [comment]
Being wasting a lot of time dealing with `stack too deep` issues... we do have to start thinking on either using ABI encoder V2 or having less optimized templates (e.g. reading settings from storage, using more txs, or sth like that). 

Note that the code looks completely awkward due to how we need to load from storage, pass by argument, and the rest of ticks and tricks that we needed to do to avoid compiling issues, it definitely won't look good for inexperienced users as an example of how to build a template :/

**Comment by ajsantander** [comment]
> I've been wasting a lot of time dealing with `stack too deep` issues... we do have to start thinking on either using ABI encoder V2 or having less optimized templates (e.g. reading settings from storage, using more txs, or sth like that).
> 
> Note that the code looks completely awkward due to how we need to load from storage, pass by argument, and the rest of ticks and tricks that we needed to do to avoid compiling issues, it definitely won't look good for inexperienced users as an example of how to build a template :/

@facuspagnuolo
ABI encoder V2 sounds nice, but we would need to either enable experimental features on contracts (not even sure if 0.4.24 has the feature or how stable it is), or update all contracts to a newer version of Solidity like 0.5.x.

IDEA: What about implementing a very simple system in which any template can be initialized with 2 transactions, no matter how simple or complex it is.

tx 1) The user deploys a "configuration contract" that simply holds state in a bunch of public variables that can be read by later by a template being initialized.

tx 2) The template's initialize function, instead of getting its settings from parameters in a function call, retrieves them from the "configuration contract" by reading state. The template could even selfdestruct the "configuration contract", getting a bit of gas refunded.

File: templates/company-board/contracts/CompanyBoardTemplate.sol [hunk 14]
```diff
@@ -47,116 +50,120 @@ contract CompanyBoardTemplate is BaseTemplate {
...
-        uint64[3] _boardVotingSettings,
-        uint64[3] _shareVotingSettings,
-        uint64 _financePeriod,
-        bool _useAgentAsVault
-    )
-        external
-    {
-        Kernel dao = _popDaoCache(msg.sender);
-        _setupApps(dao, _boardMembers, _shareHolders, _shareStakes, _boardVotingSettings, _shareVotingSettings, _financePeriod, _useAgentAsVault);
+    function setupShare(string _id, address[] _holders, uint256[] _stakes, uint64[3] _votingSettings, uint64 _financePeriod, bool _useAgentAsVault) external {
```
Comment (sohkai) [inline templates/company-board/contracts/CompanyBoardTemplate.sol]: If I understand it correctly now, the flow is:

1. `prepareInstance()`
1. `setupBoard()`
1. `setupShare()`

Is it not possible to combine `prepareInstance()` with `setupBoard()`?

File: templates/trust/package.json (hunk 8 above)
Comment (sohkai) [reply templates/trust/package.json]: Hmm, I see, the contract resolution algorithm is a bit tricky with how npm works (and it's why we can't hoist contracts when using lerna).

Not sure if truffle@5 has made any inroads to fixing this, but for now I've moved the unused apps into the dev dependencies.

**Comment by sohkai** [comment]
> using ABI encoder V2

Would ABI encoder V2 help with the stack issues? There's probably not too much we can do on that regard since we only get 16 slots?

> IDEA: What about implementing a very simple system in which any template can be initialized with 2 transactions, no matter how basic or complex it is.

I'm not sure if this would be cheaper than just passing along arguments; the main benefit I see with deploying a configuration contract (or writing to a global, shared one), would be ease of recovering if you only sent 1 / 3 transactions during onboarding or etc.

**Comment by sohkai** [comment]
@facuspagnuolo I've pushed a few commits that are mostly cosmetic / small fixes. Looking into changing the initialization structure of the company with board template (https://github.com/aragon/dao-templates/pull/124#discussion_r311472629) to see if I can come up with something more intuitive :).

File: templates/company-board/test/company-board.js [hunk 15]
```diff
@@ -21,10 +21,14 @@ const MiniMeToken = artifacts.require('MiniMeToken')
 const PublicResolver = artifacts.require('PublicResolver')
 const EVMScriptRegistry = artifacts.require('EVMScriptRegistry')
 
+const ONE_DAY = 60 * 60 * 24
+const ONE_WEEK = ONE_DAY * 7
+const THIRTY_DAYS = ONE_DAY * 30
+const TWO_MONTHS = ONE_DAY * 31
```
[the remark this answers is not in the document]
Comment (facuspagnuolo) [reply templates/company-board/test/company-board.js]: Ohhh, it definitely should be `ONE_MONTH`

File: templates/company-board/contracts/CompanyBoardTemplate.sol [hunk 16]
```diff
@@ -47,116 +50,120 @@ contract CompanyBoardTemplate is BaseTemplate {
...
+
+    function _installVaultAndFinance(uint64 _financePeriod, bool _useAgentAsVault) internal returns (Vault agentOrVault, Finance finance) {
+        Kernel dao = _fetchDaoCache();
+        agentOrVault = _useAgentAsVault ? _installDefaultAgentApp(dao) : _installVaultApp(dao);
+        finance = _installFinanceApp(dao, agentOrVault, _financePeriod == 0 ? DEFAULT_FINANCE_PERIOD : _financePeriod);
+    }
+
+    function _setupPermissions(Voting _shareVoting, TokenManager _shareTokenManager, Vault _agentOrVault, Finance _finance, bool _useAgentAsVault) internal returns (Voting) {
+        Kernel dao = _fetchDaoCache();
+        ACL _acl = ACL(dao.acl());
```
[the remark this answers is not in the document]
Comment (facuspagnuolo) [reply templates/company-board/contracts/CompanyBoardTemplate.sol]: yup

**Comment by ajsantander** [comment]
I made a quick experiment in remix to see if the configuration contract would help, and it looks like the technique produces stack too deep errors even sooner than when using plain old function parameters. If you want to have a look: https://remix.ethereum.org/#version=soljson-v0.5.1+commit.c8a2cb62.js&optimize=false&gist=17461f2c1dbd9edb598acba1a46ea412

**Comment by sohkai** [comment]
That is definitely not a good sign for the configuration contract... it's probably because doing a call involves quite a lot of stack variables.

File: templates/company-board/contracts/CompanyBoardTemplate.sol (hunk 14 above)
Comment (facuspagnuolo) [reply templates/company-board/contracts/CompanyBoardTemplate.sol]: It may be possible since they cost 5e6 and 1.3e6 respectively, which means they should fit in a single tx
